### ISSUE TITLE
BioShock 2 [NPUB30392] Constant freezing

### ISSUE THREAD

**Issue by Shoegzer** [RPCS3/rpcs3#11391]
## Quick summary
Constantly freezes while ingame.  Freezing can occur anywhere from 5-15 minutes into gameplay and seems to occur no matter where the player is in the game.

## Details
Given the "Compiling shaders..." message usually appears at the bottom whenever it freezes, guessing this is related to shader compilation threading.   However, frequency of occurrence does not seem to lessen as the game progresses and more shaders are compiled.

Tested not only default settings but also nearly all relevant settings in CPU, GPU and Advanced tabs to no positive effect.

## System configuration
* OS: Linux Mint 20.3 / kernel 5.13
* CPU: Intel i7-10870H
* GPU: NVidia GTX 3070 / driver 470.86
* Driver version: 470.86
* Commit tested: 5b980b99a03cebe11784f1caf93e0337a4126571
* Logfile: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/7878743/RPCS3.log.gz)
(Note: removed large repeating section in middle of log since original logfile was over 150mb, making viewing the file difficult at least on this test system).

**Comment by kd-11** [comment]
Please install a more recent driver. Nvidia's 47x series is riddled with bugs.

**Comment by Shoegzer** [comment]
Thanks @kd-11.  I installed 495.46 and tried again but with same results unfortunately.  Updated issue with this info.

**Comment by kd-11** [comment]
Hi, is it possible to get a vulkan sdk trace? Just enable 'debug output' with the sdk installed and attach the log when it crashes.
While this issue has been reported widely, we're having trouble reproducing it.

**Comment by Shoegzer** [comment]
You bet.  Added the logfile with sdk trace to the issue post above.  Hope it helps, but just let me know if you need anything else and I'm happy to help.

**Comment by kd-11** [comment]
This was a big help. Thanks.
```
·E 0:18:17.068386 {RSX [0x000abe4]} RSX: ERROR: [Validation] Code 0 : Validation Error: [ VUID-vkCreateSampler-maxSamplerAllocationCount-04110 ] Object 0: handle = 0x7fdea1f90650, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xb408bc0b | vkCreateSampler(): Number of currently valid sampler objects (4092) is not less than the maximum allowed (4000). The Vulkan spec states: There must be less than VkPhysicalDeviceLimits::maxSamplerAllocationCount VkSampler objects currently created on the device. (https://vulkan.lunarg.com/doc/view/1.2.170.0~rc2/linux/1.2-extensions/vkspec.html#VUID-vkCreateSampler-maxSamplerAllocationCount-04110)
```
This is the reason for the crash, the sampler pool ran out which means we're likely leaking sampler objects. This also explains why it crashes at regular intervals. I'll push out a fix soon.

**Comment by kd-11** [comment]
Should be fixed by https://github.com/RPCS3/rpcs3/pull/11401
Please test.

**Comment by Shoegzer** [comment]
Thanks, I'm on it.  When your PR is done building I'll test for an hour, that should be plenty of time to confirm based on prior occurrences.  Results to follow.

Incidentally, do you know why the game froze during my opengl tests, or will your new sampler pool limit resolve that issue too?

**Comment by kd-11** [comment]
No, the OGL freeze is likely a completely separate thing. I'll look at it once we have an answer for the nvidia vulkan issues.

**Comment by Shoegzer** [comment]
Thanks and that makes sense, it just seemed odd that the same issue would be triggered by a different condition.  Hopefully this can at least lend clues.

**Comment by Shoegzer** [comment]
...and we're good, over an hour ingame with no freezing!  The sampler level was, you know, over 4000.  Who knew?  Thanks so much.

I suppose I should keep this open due to freezing on the opengl side but again happy to help you test anything further.

**Comment by Shoegzer** [comment]
Would you advise opening a new issue to track the opengl freezing?  If so, please let me know if you need anything for it beyond a basic log.  Thanks.

**Comment by kd-11** [comment]
> Would you advise opening a new issue to track the opengl freezing? If so, please let me know if you need anything for it beyond a basic log. Thanks.

Yes, lets get a new ticket opened for that. I was not able to get any freezing with OpenGL other than just performance hitching.

**Comment by Shoegzer** [comment]
I encountered another crash in this game with default settings and vulkan, though based on the logs it looks like this may not be related to the sampler pool.  It was a one-time occurrence and may even be related to something you're aware of, so please feel free to disregard, but I felt you should know.

* Commit: 2f7d38bb8178e9cf75727c01b675343011a2326e
* Logfile: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/7937794/RPCS3.log.gz)

**Comment by kd-11** [comment]
This is unrelated and seems to be a PPU crash. Nothing to do with rendering.
If you find a way to make it happen frequently enough, open a ticket for it and someone will investigate.
